client/LaunchVM.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) 2022 IBM
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+import base64
+import libvirt
+import shutil
+import grpc
+import hashlib
+from uuid import UUID
+from bs4 import BeautifulSoup as bs
+
+from keybroker_pb2_grpc import KeyBrokerServiceStub
+from keybroker_pb2 import BundleRequest, SecretRequest, RequestDetails
+
+XML_PATH = "sev_guest.xml"
+KBS_URI = "127.0.0.1:44444"
+
+def main():
+    hv = libvirt.open("qemu:///system")
+
+    # load and parse XML
+    f = open(XML_PATH, "r")
+    xml = f.read()
+    f.close()
+
+    soup = bs(xml, "xml")
+    
+    policy = soup.find("policy").text
+
+    # use this to calculate the expected digest
+    ovmf_path = soup.find("loader").text
+    initrd_path = soup.find("initrd").text
+    kernel_path = soup.find("kernel").text
+    cmdline = soup.find("cmdline").text
+
+    channel = grpc.insecure_channel(KBS_URI)
+    client = KeyBrokerServiceStub(channel)
+   
+    
+    # currrently libvirt cannot export a cert chain with a 
+    # signed CEK. instead, use sevctl to export the cert chain
+    # and store it in a file.
+    f = open("certchain","rb")
+    cert_chain_bytes = f.read()
+    f.close()
+    cert_chain = base64.b64encode(cert_chain_bytes).decode("utf-8")
+
+    request = BundleRequest(CertificateChain = cert_chain, Policy = int(policy, 0))
+    response = client.GetBundle(request)
+
+    # add godh and session file to XML
+    ls = soup.find("launchSecurity")
+    cert_tag = soup.new_tag("dhCert")
+    cert_tag.string = response.GuestOwnerPublicKey
+    ls.append(cert_tag)
+
+    session_tag = soup.new_tag("session")
+    session_tag.string = response.LaunchBlob
+    ls.append(session_tag)
+
+    xml = str(soup.domain)
+
+    # define the domain from XML
+    dom = hv.defineXML(xml) 
+    dom.createWithFlags(libvirt.VIR_DOMAIN_START_PAUSED)
+
+    # get launch measurement
+    sevinfo = dom.launchSecurityInfo()
+
+    expected_digest = get_expected_digest(ovmf_path, initrd_path, kernel_path, cmdline)
+
+    request = SecretRequest(LaunchMeasurement = sevinfo['sev-measurement'], \
+            LaunchId = response.LaunchId, \
+            Policy = sevinfo['sev-policy'], \
+            ApiMajor = sevinfo['sev-api-major'], \
+            ApiMinor = sevinfo['sev-api-minor'], \
+            BuildId = sevinfo['sev-build-id'], \
+            FwDigest = expected_digest, \
+            LaunchDescription = "test launch",
+            SecretRequests = [\
+                    RequestDetails(Guid = "0a46e24d-478c-4eb1-8696-113eeec3aa99", \
+                        Format = "json", \
+                        SecretType = "bundle", \
+                        Id = "keyset1" )] \
+            )
+   
+    response = client.GetSecret(request)
```

Comment:
Shouldn't you deal a bit more gracefully with possible errors here? Or is a python stack trace good enough (and providing enough info about what went wrong)?

Reply:
Yeah, I can add some error handling here. This LaunchVM script was mainly for testing. It is not very sophisticated.